src/Elements/types.jl

```diff
@@ -0,0 +1,180 @@
+                    :typealias, :global,   :symbol, :tuple, :aside]
+
+
+## Config Nodes
+abstract ConfigNode
+
+type Config <: ConfigNode
+    objfilter               :: Any
+    style_obj               :: ASCIIString
+
+    const defaults = Dict{Symbol, Any}(
+        :objfilter                  => CATEGORIES,
+        :style_obj                  => "####",
+        )
+
+    Config(; args...) = update_config!(new(), merge(defaults, Dict(args)))
+end
+config(; cargs...) = PreConfig(; cargs...)
+
+type PreConfig <: ConfigNode
+    cargs :: Dict{Symbol, Any}
+
+    PreConfig(; cargs...) = new(Dict(cargs))
+end
+
+function update_config!(config::Config, args::Dict)
+    const fields = fieldnames(Config)
+    for (k, v) in args
+        k in fields                                                ?
+            setfield!(config, k, convert(fieldtype(Config, k), v)) :
+            throw(ArgumentError("'Config'. Invalid setting: '$(k) = $(v)'."))
+    end
+
+    for k in [:style_obj]
+        getfield(config, k) in HEADERS ||
+                error("""Invalid style : config-item `$k -> $(getfield(config, k))`.
+                      Valid values: [$(join(MDSTYLETAGS, ", "))].""")
+    end
+    return config
+end
+update_config!(config::Config, args...) = update_config!(config, Dict(args))
+
+
+## Main structure nodes
+abstract SectionNode
+abstract PageNode
+abstract ContentNode
+
+immutable NullSection <: SectionNode end
+immutable NullPage    <: PageNode    end
+
+# Document
+type Document
+    name        :: AbstractString
+    config      :: ConfigNode
+    meta        :: Dict{Symbol, Any}
+    children    :: Vector{SectionNode}
+end
+document(name::AbstractString, conf::PreConfig, children...) =
+        document(name, conf, Dict(), [children...])
+document(name::AbstractString, children...) = document(name, PreConfig(), children...)
+
+function document(name::AbstractString, conf::PreConfig, meta::Dict, children::Vector)
+    this = isempty(conf.cargs)                       ?
+            Document(name, Config(), meta, children) :
+            Document(name, update_config!(Config(), deepcopy(conf.cargs)), meta, children)
+    isempty(children) || postprocess!(this, children)
+    return this
+end
+
+# Section
+type Section <: SectionNode
+    name            :: AbstractString
+    config          :: ConfigNode
+    meta            :: Dict{Symbol, Any}
+    parent          :: Union(Document, SectionNode)
+    children        :: Vector{Union(SectionNode, PageNode)}
+end
+section(name::AbstractString, conf::PreConfig, children...) =
+        Section(name, conf, Dict(:autogenerate => (false, nothing)), NullSection(), [children...])
+section(name::AbstractString, children...) = section(name, PreConfig(), children...)
+
+# Page
+type Page <: PageNode
+    name            :: AbstractString
+    config          :: ConfigNode
+    meta            :: Dict{Symbol, Any}
+    parent          :: SectionNode
+    children        :: Vector{ContentNode}
+end
```

Comment:
``Document``, ``Section``, and ``Page`` all appear to have the same structure. Perhaps they should just be

```jl
abstract NodeT

immutable Document <: NodeT end
immutable Section <: NodeT end
immutable Page <: NodeT end

type Node{T <: NodeT}
   name
   config
   meta
   parent
   children
end
```

``parent`` for a ``Document`` would just be a self reference. Should ``name`` just be a field in the ``meta`` ``Dict``?